### PR TITLE
Site Groups pecan 629

### PR DESCRIPTION
**Pull request by robkooper**
This contains the code to add a migration and interface to add site groups. This is a group of sites that logically belong together.

### CODE REVIEW

**Comment by gsrohde** [comment]
Still some references to "cluster" (see line notes).

File: app/models/sitegroup.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class Sitegroup < ActiveRecord::Base
+  include Overrides
+
+  extend SimpleSearch
+  SEARCH_INCLUDES = %w{ }
+  SEARCH_FIELDS = %w{ clusters.name }
```
Comment (gsrohde) [inline app/models/sitegroup.rb]: --> sitegroups.name

File: app/views/sitegroups/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<div class="content">
+	<div class="container">
+		<div class="sixteen columns">
+			<header>
+	      <h1>Editing Cluster</h1>
```
Comment (gsrohde) [inline app/views/sitegroups/edit.html.erb]: --> Editing Site Group

File: db/migrate/20160412030352_add_sitegroups.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      t.string :name, null: false
+      t.boolean :public_access, null: false, :default => false
+      t.integer :user_id, :limit => 8, null: false
+      t.timestamps
+    end
+    change_column :sitegroups, :id, :integer, :limit => 8
+    # fix the counter for inserting new records
+    execute %{
+        SELECT setval('sitegroups_id_seq', GREATEST(1, CAST(1e9 * #{this_hostid}::int AS bigint)), FALSE);
+        ALTER TABLE clusters
```
Comment (gsrohde) [inline db/migrate/20160412030352_add_sitegroups.rb]: --> ALTER TABLE sitesgroups

File: db/migrate/20160412030352_add_sitegroups.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        SELECT setval('sitegroups_sites_id_seq', GREATEST(1, CAST(1e9 * #{this_hostid}::int AS bigint)), FALSE);
+        ALTER TABLE sitegroups_sites
+            ALTER COLUMN created_at SET DEFAULT utc_now(),
+            ALTER COLUMN updated_at SET DEFAULT utc_now();
+
+        ALTER TABLE "sitegroups_sites"
+            ADD CONSTRAINT "fk_sitegroups_sites_sites"
+                FOREIGN KEY ("site_id") REFERENCES "sites" ("id")
+                ON DELETE CASCADE ON UPDATE CASCADE,
+            ADD CONSTRAINT "fk_sitegroups_sites_clusters"
```
Comment (gsrohde) [inline db/migrate/20160412030352_add_sitegroups.rb]: --> "fk_sitegroups_sites_sitegroups"

File: db/migrate/20160412030352_add_sitegroups.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                FOREIGN KEY ("sitegroup_id") REFERENCES "clusters" ("id")
```
Comment (gsrohde) [inline db/migrate/20160412030352_add_sitegroups.rb]: --> REFERENCES "sitegroups"

File: db/migrate/20160412030352_add_sitegroups.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                ON DELETE CASCADE ON UPDATE CASCADE;
+      }
+
+    # only execute if migration is running on host 0
+    if this_hostid == 0
+      execute %{
+        INSERT INTO clusters (name, everybody, user_id)
```
Comment (gsrohde) [inline db/migrate/20160412030352_add_sitegroups.rb]: --> INTO sitegroups

File: db/migrate/20160412030352_add_sitegroups.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            VALUES('AmeriFlux', true, 0);
+        INSERT INTO sitegroups_sites (sitegroup_id, site_id)
+            SELECT (SELECT id FROM clusters WHERE name='AmeriFlux'), id
```
Comment (gsrohde) [inline db/migrate/20160412030352_add_sitegroups.rb]: --> FROM sitegroups

File: spec/features/sitegroup_integration_spec.rb [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      @private_sitegroups_of_other_users.each do |c|
+        visit "/sitegroups/#{c.id}"
+        expect(page).not_to have_content 'Viewing Sitegroup'
+      end
+    end
+
+    specify "should be able to view private sitegroups of his own" do
+      @private_sitegroups_of_this_user.each do |c|
+        visit "/sitegroups/#{c.id}"
+        expect(page).to have_content 'Viewing Sitegroup'
```
Comment (gsrohde) [inline spec/features/sitegroup_integration_spec.rb]: 'Viewing Sitegroup' --> 'Viewing Site Group'

File: spec/features/sitegroup_integration_spec.rb [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      visit '/sitegroups'
+      expect(page).not_to have_xpath('.//tr[td/text() = "public_sitegroup"][td/a[@data-method = "delete"]]')
+    end
+
+    specify "shouldn't see an delete button even for a public sitegroup created by himself" do
+      visit '/sitegroups'
+      expect(page).not_to have_xpath('.//tr[td/text() = "creator_public_sitegroup"][td/a[@data-method = "delete"]]')
+    end
+
+    specify "shouldn't have edit access to even a public sitegroup created by another user" do
```
Comment (gsrohde) [inline spec/features/sitegroup_integration_spec.rb]: Change 
```
specify "shouldn't have edit access to even a public sitegroup created by another user" do
```
 to
```
specify "shouldn't have edit access to even a public sitegroup created by another user",
pending: "Restricting access to the edit pages for non-administrators" do
```

File: spec/features/sitegroup_integration_spec.rb [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+      visit "/sitegroups/#{public_sitegroup.id}/edit"
+      expect(page).not_to have_content 'Editing Sitegroup'
```
Comment (gsrohde) [inline spec/features/sitegroup_integration_spec.rb]: 'Editing Sitegroup' --> 'Editing Site Group'

File: spec/features/sitegroup_integration_spec.rb [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    specify "should see delete buttons for all items in the sitegroups list" do
+      visit '/sitegroups'
+      expect(all(:xpath, ".//tbody/tr[td/a[@data-method = 'delete']]").length).to eq Sitegroup.count
+    end
+
+    specify "should be able to view private sitegroups created by other users" do
+      sitegroup = sitegroups(:private_sitegroup_from_creator_user)
+      id = sitegroup.id
+      visit "/sitegroups/#{id}"
+      expect(page).to have_content "Viewing Sitegroup"
```
Comment (gsrohde) [inline spec/features/sitegroup_integration_spec.rb]: 'Viewing Sitegroup' --> 'Viewing Site Group'

File: spec/features/sitegroup_integration_spec.rb [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+      expect(page).to have_content sitegroup.name
+    end
+
+    specify "should have edit access to private sitegroups created by other users" do
+      sitegroup = sitegroups(:private_sitegroup_from_creator_user)
+      id = sitegroup.id
+      visit "/sitegroups/#{id}/edit"
+      expect(page).to have_content "Editing Sitegroup"
```
Comment (gsrohde) [inline spec/features/sitegroup_integration_spec.rb]: 'Editing Sitegroup' --> 'Editing Site Group'

File: test/fixtures/users.yml [hunk 13]
```diff
@@ -28,30 +30,35 @@ test_user2:
 
 test_user3:
     login: 'robben_yang'
+    name 'Robben Yang'
```
Comment (gsrohde) [inline test/fixtures/users.yml]: I left out a colon here.  Should be
```
    name: 'Robben Yang'
```

**Comment by robkooper** [comment]
finally got it to pass all the tests

**Comment by gsrohde** [comment]
Will try to release today.

**Pull request merged**